src/RA_Core.cpp

```diff
@@ -56,10 +56,23 @@ inline static unsigned int g_nProcessTimer{ 0U };
 
 BOOL APIENTRY DllMain(HMODULE hModule, DWORD dwReason, _UNUSED LPVOID)
 {
-    if (dwReason == DLL_PROCESS_ATTACH)
-        g_hThisDLLInst = hModule;
+    // exceptions can happen here, they must not be allowed to leave
+    try
+    {
+        if (dwReason == DLL_PROCESS_ATTACH)
+        {
+            // resetting it more than once will cause a deadlock or access violation
+            // Library will be freed when the app ends
+            if (!g_hThisDLLInst)
+                g_hThisDLLInst.reset(hModule);
```

Comment:
This is really important, it can only be assigned once since it's a unique_ptr. I'd personally prefer not to deal with cycles, `std::shared_ptr` is potentially slow (not always but definitely slower than `std::unique_ptr` if you don't use `std::weak_ptr` to break the cycles). Plus `std::shared_ptr` doesn't support custom deleters.

Reply:
Library should be freed by the client, not the library. We are simply caching our own base pointer (see https://docs.microsoft.com/en-us/windows/desktop/dlls/dllmain)
> A handle to the DLL module. The value is the base address of the DLL. 

Deleting ourself, even if it is when we're being unloaded, can lead to undefined behavior.